packages/api/internal/dns/server.go

```diff
@@ -2,99 +2,242 @@ package dns
 	"fmt"
-	"log"
 	"net"
 	"strings"
 	"sync"
+	"time"
 
+	"github.com/go-redis/cache/v8"
 	"github.com/go-redis/redis/v8"
 	resolver "github.com/miekg/dns"
+	"go.uber.org/zap"
 
 	"github.com/e2b-dev/infra/packages/shared/pkg/smap"
 )
 
 const ttl = 0
+const redisTTL = 24 * time.Hour
 
 const defaultRoutingIP = "127.0.0.1"
 
+const cachedDnsPrefix = "sandbox.dns."
+
 type DNS struct {
-	mu      sync.Mutex
-	redis   *redis.Client
-	records *smap.Map[string]
+	srv    *resolver.Server
+	logger *zap.SugaredLogger
+
+	remote *cache.Cache
+	local  *smap.Map[string]
+
+	closer struct {
+		once sync.Once
+		op   func(context.Context) error
+		err  error
+	}
 }
 
-func New(rc *redis.Client) *DNS {
-	return &DNS{
-		redis:   rc,
-		records: smap.New[string](),
+func New(ctx context.Context, rc *redis.Client, logger *zap.SugaredLogger) *DNS {
+	d := &DNS{logger: logger}
+
+	if rc != nil {
+		d.remote = cache.New(&cache.Options{Redis: rc, LocalCache: cache.NewTinyLFU(10_000, time.Hour)})
+	} else {
+		d.local = smap.New[string]()
 	}
+
+	return d
 }
 
-func (d *DNS) Add(_ context.Context, sandboxID, ip string) {
-	d.records.Insert(d.hostname(sandboxID), ip)
+func (d *DNS) Add(ctx context.Context, sandboxID, ip string) {
+	switch {
+	case d.remote != nil:
+		d.remote.Set(&cache.Item{
+			Ctx:   ctx,
+			TTL:   redisTTL,
+			Key:   sandboxID,
+			Value: ip,
+		})
+	case d.local != nil:
+		d.local.Insert(sandboxID, ip)
+	}
 }
 
-func (d *DNS) Remove(_ context.Context, sandboxID, ip string) {
-	d.records.RemoveCb(d.hostname(sandboxID), func(key string, v string, exists bool) bool {
-		return v == ip
-	})
+func (d *DNS) Remove(ctx context.Context, sandboxID, ip string) {
+	switch {
+	case d.remote != nil:
+		if err := d.remote.Delete(ctx, d.cacheKey(sandboxID)); err != nil {
+			d.logger.Debug("removing item from DNS cache", zap.Error(err), zap.String("sandbox", sandboxID))
+		}
+	case d.local != nil:
+		d.local.RemoveCb(d.cacheKey(sandboxID), func(k string, v string, ok bool) bool { return v == ip })
+	}
 }
 
-func (d *DNS) get(hostname string) (string, bool) {
-	return d.records.Get(hostname)
+func (d *DNS) Get(ctx context.Context, sandboxID string) net.IP {
+	var res string
+	switch {
+	case d.remote != nil:
+		if err := d.remote.Get(ctx, d.cacheKey(sandboxID), &res); err != nil {
+			if errors.Is(err, cache.ErrCacheMiss) {
+				d.logger.Warn("item missing in remote DNS cache", zap.String("sandbox", sandboxID))
+			} else {
+				d.logger.Error("resolving item from remote DNS cache", zap.String("sandbox", sandboxID), zap.Error(err))
+			}
+		}
+	case d.local != nil:
+		var ok bool
+		res, ok = d.local.Get(d.cacheKey(sandboxID))
+		if !ok {
+			d.logger.Warn("item not found in local DNS cache", zap.String("sandbox", sandboxID))
+		}
+	}
+
+	addr := net.ParseIP(res)
+	if addr == nil {
+		if res != "" {
+			d.logger.Error("malformed address in cache", zap.Bool("local", d.local != nil), zap.String("addr", res))
+		}
+
+		addr = net.ParseIP(defaultRoutingIP)
```

Comment:
I put the "fall back to localhost" in the Get method from the cache, where it was previously in the handler. 

This behavior is weird (note: the DNS service doesn't run in when in "local mode" so any time we return a local address it may be wrong.)